### DISCUSSION TITLE
Login on another page/server

### DISCUSSION THREAD

**Discussion by sergioocode** [reactiveui/refit#1561] [General]
Hello, in this moment i am using "IHttpClientFactory" and all works fine.
I need to know if is possible this approach:

` private const string clientName = "SOC";`

-login auth (with cert x509)
`public async Task Connect(string url)
{
    HttpClient httpclient = clientFactory.CreateClient(clientName);
    var res = await httpclient!.GetAsync(
        Path.Combine(
            string.Format(
                "https://serverlogin.com/cgi_AUT2000/CAutInicio.cgi?reference={0}",
                url
            )
        )
    );
}`

-Get info
`public async Task GetInfo(HttpRequestMessage message)
{
    message.RequestUri = new Uri("https://server2.com/cve_cgi/api/users/");
    HttpClient httpclient = clientFactory.CreateClient(clientName);
    HttpResponseMessage? res = await httpclient!.SendAsync(message);
}`

This is DI:
`services
    .AddHttpClient(
        "SOC",
        c =>
        {
            c.DefaultRequestHeaders.UserAgent.ParseAdd("Mozilla/5.0");
        }
    )
    .ConfigurePrimaryHttpMessageHandler(() =>
    {
        CookieContainer? cookieContainer = new();
        HttpClientHandler? handler =
            new() { CookieContainer = cookieContainer };
        handler.ClientCertificates.Add(
            DigitalCert.GetCert(context.Configuration)
        );
        return handler;
    });`

This methods works fine.

So, now i want use the Refit library. Is this possible? 
I have problems because i dont knows how change the base address...

` services
     .AddRefitClient<IInterface>()
     .ConfigureHttpClient(c =>
     {
         c.DefaultRequestHeaders.UserAgent.ParseAdd("Mozilla/5.0");
         c.BaseAddress = new Uri("https://server2.com/cve_cgi/api/users/");
     })
     .ConfigurePrimaryHttpMessageHandler(() =>
     {
         CookieContainer? cookieContainer = new();
         HttpClientHandler? handler =
             new() { CookieContainer = cookieContainer };
         handler.ClientCertificates.Add(
             DigitalCert.GetCert(context.Configuration)
         );
         return handler;
     });`

**Comment by glennawatson** [comment]
Yes, this works with Refit. The key points:

1. Base address: set it per Refit client via `ConfigureHttpClient(c => c.BaseAddress = ...)`, exactly as you already did. Each Refit interface can target its own host. If you need two different hosts (login server vs API server), register two Refit clients.

2. Shared cookies (the login flow): the cookie set by the login call must be visible to the API call. Share a single `CookieContainer` across both handlers. The cleanest way is `ConfigurePrimaryHttpMessageHandler` returning a handler that uses the same `CookieContainer` instance for both clients:

```csharp
var cookies = new CookieContainer();

Func<HttpMessageHandler> handlerFactory = () =>
{
    var handler = new HttpClientHandler { CookieContainer = cookies, UseCookies = true };
    handler.ClientCertificates.Add(DigitalCert.GetCert(context.Configuration));
    return handler;
};

services.AddRefitClient<ILoginApi>()
    .ConfigureHttpClient(c => c.BaseAddress = new Uri("https://serverlogin.com/"))
    .ConfigurePrimaryHttpMessageHandler(handlerFactory);

services.AddRefitClient<IUsersApi>()
    .ConfigureHttpClient(c => c.BaseAddress = new Uri("https://server2.com/cve_cgi/api/users/"))
    .ConfigurePrimaryHttpMessageHandler(handlerFactory);
```

3. Define the calls as Refit methods:

```csharp
public interface ILoginApi
{
    [Get("/cgi_AUT2000/CAutInicio.cgi")]
    Task Connect([Query] string reference);
}

public interface IUsersApi
{
    [Get("")]
    Task<UserInfo> GetInfo();
}
```

Call `Connect` first so the auth cookie lands in the shared `CookieContainer`, then call `GetInfo` on the other client; the cookie flows automatically.

Notes:
- The x509 client cert goes on the primary handler, as you have it.
- If both hosts share a cookie domain, one shared `CookieContainer` is enough. If they are unrelated domains and the cookie is host-scoped, the cookie will only be sent to the host that set it - in that case carry the token/cookie forward manually via a `DelegatingHandler`.
- You generally do not need to construct `HttpRequestMessage` yourself with Refit; let the interface methods build the request.
